src/utils/provenance.js

```diff
@@ -0,0 +1,70 @@
+'use strict'
+const semver = require('semver')
+const { execWithOutput } = require('./execWithOutput')
+
+/**
+ * Abort if the user specified they want NPM provenance, but their CI's NPM version doesn't support it.
+ * If we continued, the release will go ahead with no warnings, and no provenance will be generated.
+ */
+function checkIsSupported(npmVersion) {
+  const validNpmVersion = '>=9.5.0'
+
+  if (!semver.satisfies(npmVersion, validNpmVersion)) {
+    throw new Error(
+      `Provenance requires NPM ${validNpmVersion}, but this action is using v${npmVersion}.
+Either remove provenance from your release action's inputs, or update your release CI's NPM version.`
+    )
+  }
+}
+
+/**
+ * Abort with a meaningful error if the user would get a misleading error message from NPM
+ * due to an NPM bug that existed between 9.5.0 and 9.6.1.
+ * As of April 2023, this would affect anyone whose CI is set to Node 18 (which defaults to NPM 9.5.1).
+ */
+function checkPermissions(npmVersion) {
+  // Bug was fixed in this NPM version - see https://github.com/npm/cli/pull/6226
+  const correctNpmErrorVersion = '>=9.6.1'
+
+  if (
+    // Same test condition as in fixed versions of NPM
+    !process.env.ACTIONS_ID_TOKEN_REQUEST_URL &&
```

Comment:
is this what tells us that we have the permissions? Q: do we really need to test against this? can't we just let it fail if the user did not configure the permission? e.g. what if GH changes the implementation and this variable is gone?

Reply:
> generally speaking I'm not a fan of programming against specific versions of tools we use, npm in this case

Agreed, me too. In this case I think it's worth making an exception because our recommended config gives users a version of NPM containing a bug that we know actively misinforms them about their problem.

> can't we just let it fail if the user did not configure the permission?

Not when NPM version < 9.6.1 (i.e. everyone using `ubuntu-latest` as per our recommended config): that linked NPM bug causes them to get a very misleading error message _"Automatic provenance generation not supported outside of GitHub Actions"_ which will send them in entirely the wrong direction. 

> what if GH changes the implementation and this variable is gone?

This check is already limited to only apply on NPM < 9.6.1 in the next line. Every NPM version > 9.5.0 includes this same check on `!process.env.ACTIONS_ID_TOKEN_REQUEST_URL`, so if GitHub did make that change, regardless of what we do it'd break provenance for everyone using any currently released version of NPM.

So all this check does is make buggy but still widely used NPM versions between 9.5.0 and 9.6.0 behave the same as the fixed version NPM 9.6.1; it'll have no effect at all on any NPM versions beyond 9.6.0, and doesn't add any brittleness to the affected versions that wasn't already there.